{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "037a7b25",
   "metadata": {},
   "source": [
    "1. 导入库与配置"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "71ddb55e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_10204/3682314905.py:8: FutureWarning: Use `scanpy.set_figure_params` instead\n",
      "  sc.settings.set_figure_params(dpi=80, facecolor='white')\n"
     ]
    }
   ],
   "source": [
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "# 设置 Scanpy 的绘图参数\n",
    "sc.settings.verbosity = 3             # 显示日志\n",
    "sc.logging.print_header()\n",
    "sc.settings.set_figure_params(dpi=80, facecolor='white')\n",
    "\n",
    "results_file = 'pbmc3k_results.h5ad'  # 结果保存路径"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13a88666",
   "metadata": {},
   "source": [
    "2. 下载并读取数据\n",
    "Scanpy 内置了下载函数，非常方便："
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "1243096e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 下载 3k PBMC 数据（来自 10x Genomics）\n",
    "# 修改为官方标准函数\n",
    "adata = sc.datasets.pbmc3k() \n",
    "adata.var_names_make_unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "f46899f6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_genes_by_counts</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>total_counts_mt</th>\n",
       "      <th>pct_counts_mt</th>\n",
       "      <th>n_genes</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACATACAACCAC-1</th>\n",
       "      <td>781</td>\n",
       "      <td>2421.0</td>\n",
       "      <td>73.0</td>\n",
       "      <td>3.015283</td>\n",
       "      <td>781</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACATTGAGCTAC-1</th>\n",
       "      <td>1352</td>\n",
       "      <td>4903.0</td>\n",
       "      <td>186.0</td>\n",
       "      <td>3.793596</td>\n",
       "      <td>1352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACATTGATCAGC-1</th>\n",
       "      <td>1131</td>\n",
       "      <td>3149.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>0.889171</td>\n",
       "      <td>1131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCGTGCTTCCG-1</th>\n",
       "      <td>960</td>\n",
       "      <td>2639.0</td>\n",
       "      <td>46.0</td>\n",
       "      <td>1.743085</td>\n",
       "      <td>960</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCGTGTATGCG-1</th>\n",
       "      <td>522</td>\n",
       "      <td>981.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>1.223242</td>\n",
       "      <td>522</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  n_genes_by_counts  total_counts  total_counts_mt  \\\n",
       "index                                                                \n",
       "AAACATACAACCAC-1                781        2421.0             73.0   \n",
       "AAACATTGAGCTAC-1               1352        4903.0            186.0   \n",
       "AAACATTGATCAGC-1               1131        3149.0             28.0   \n",
       "AAACCGTGCTTCCG-1                960        2639.0             46.0   \n",
       "AAACCGTGTATGCG-1                522         981.0             12.0   \n",
       "\n",
       "                  pct_counts_mt  n_genes  \n",
       "index                                     \n",
       "AAACATACAACCAC-1       3.015283      781  \n",
       "AAACATTGAGCTAC-1       3.793596     1352  \n",
       "AAACATTGATCAGC-1       0.889171     1131  \n",
       "AAACCGTGCTTCCG-1       1.743085      960  \n",
       "AAACCGTGTATGCG-1       1.223242      522  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs.head() # 查看结果"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c3091923",
   "metadata": {},
   "source": [
    "3. 预处理与质控 (QC)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "26abdd8b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_10204/2406383915.py:11: ImplicitModificationWarning: Trying to modify attribute `.var` of view, initializing view as actual.\n",
      "  adata.var['mt'] = adata.var_names.str.startswith('MT-')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_genes_by_counts</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>total_counts_mt</th>\n",
       "      <th>pct_counts_mt</th>\n",
       "      <th>n_genes</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACATACAACCAC-1</th>\n",
       "      <td>779</td>\n",
       "      <td>2419.0</td>\n",
       "      <td>73.0</td>\n",
       "      <td>3.017776</td>\n",
       "      <td>781</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACATTGAGCTAC-1</th>\n",
       "      <td>1352</td>\n",
       "      <td>4903.0</td>\n",
       "      <td>186.0</td>\n",
       "      <td>3.793596</td>\n",
       "      <td>1352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACATTGATCAGC-1</th>\n",
       "      <td>1129</td>\n",
       "      <td>3147.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>0.889736</td>\n",
       "      <td>1131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCGTGCTTCCG-1</th>\n",
       "      <td>960</td>\n",
       "      <td>2639.0</td>\n",
       "      <td>46.0</td>\n",
       "      <td>1.743085</td>\n",
       "      <td>960</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCGTGTATGCG-1</th>\n",
       "      <td>521</td>\n",
       "      <td>980.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>1.224490</td>\n",
       "      <td>522</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  n_genes_by_counts  total_counts  total_counts_mt  \\\n",
       "index                                                                \n",
       "AAACATACAACCAC-1                779        2419.0             73.0   \n",
       "AAACATTGAGCTAC-1               1352        4903.0            186.0   \n",
       "AAACATTGATCAGC-1               1129        3147.0             28.0   \n",
       "AAACCGTGCTTCCG-1                960        2639.0             46.0   \n",
       "AAACCGTGTATGCG-1                521         980.0             12.0   \n",
       "\n",
       "                  pct_counts_mt  n_genes  \n",
       "index                                     \n",
       "AAACATACAACCAC-1       3.017776      781  \n",
       "AAACATTGAGCTAC-1       3.793596     1352  \n",
       "AAACATTGATCAGC-1       0.889736     1131  \n",
       "AAACCGTGCTTCCG-1       1.743085      960  \n",
       "AAACCGTGTATGCG-1       1.224490      522  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 计算线粒体基因比例， pp 是 Pre-Processing（预处理）的缩写\n",
    "# sc.pp (Pre-processing)：数据清洗、标准化、质控、去噪。\n",
    "# sc.tl (Tools)：更高级的分析，如聚类、降维（UMAP）、轨迹追踪。\n",
    "# sc.pl (Plotting)：绘图和可视化。\n",
    "# percent_top=None含义：是否计算前 N 个高表达基因所占的比例。作用：通常设为 None 或 [50, 100, 200, 500]。\n",
    "# 在单细胞中，如果前 50 个基因就占了总表达量的 80%，说明这个细胞的多样性极低，可能是文库质量问题。\n",
    "# log1p=False含义：是否对计算出的指标进行 $log(x+1)$ 处理。作用：这里设为 False，我们通常直接看原始比例和计数。\n",
    "# inplace=True含义：是否直接修改原对象。作用：设为 True，意味着该函数没有返回值，而是直接更新你的 adata 变量。\n",
    "# 这能节省内存（这对你保持 C 盘空间也有好处）。\n",
    "\n",
    "adata.var['mt'] = adata.var_names.str.startswith('MT-')\n",
    "sc.pp.calculate_qc_metrics(adata, qc_vars=['mt'], percent_top=None, log1p=False, inplace=True)\n",
    "adata.obs.head() # 查看结果，你会发现多了几列：\n",
    "# n_genes_by_counts：每个细胞检测到了多少个基因（基因丰度）。\n",
    "# total_counts：每个细胞的总 UMI 计数（测序深度）。\n",
    "# pct_counts_mt：最重要的指标——线粒体基因占比。\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "7d699d7f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "filtered out 17 genes that are detected in less than 3 cells\n"
     ]
    },
    {
     "ename": "TypeError",
     "evalue": "cannot do positional indexing on Index with these indexers [50] of type str",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mTypeError\u001b[39m                                 Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[18]\u001b[39m\u001b[32m, line 5\u001b[39m\n\u001b[32m      3\u001b[39m sc.pp.filter_genes(adata, min_cells=\u001b[32m3\u001b[39m)\n\u001b[32m      4\u001b[39m adata = adata[adata.obs.pct_counts_mt < \u001b[32m5\u001b[39m, :] \u001b[38;5;66;03m# 过滤线粒体占比高的细胞\u001b[39;00m\n\u001b[32m----> \u001b[39m\u001b[32m5\u001b[39m \u001b[43madata\u001b[49m\u001b[43m.\u001b[49m\u001b[43mobs\u001b[49m\u001b[43m.\u001b[49m\u001b[43mhead\u001b[49m\u001b[43m(\u001b[49m\u001b[33;43m'\u001b[39;49m\u001b[33;43m50\u001b[39;49m\u001b[33;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/miniforge3/envs/scanpy_env/lib/python3.12/site-packages/pandas/core/generic.py:5761\u001b[39m, in \u001b[36mNDFrame.head\u001b[39m\u001b[34m(self, n)\u001b[39m\n\u001b[32m   5674\u001b[39m \u001b[38;5;129m@final\u001b[39m\n\u001b[32m   5675\u001b[39m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34mhead\u001b[39m(\u001b[38;5;28mself\u001b[39m, n: \u001b[38;5;28mint\u001b[39m = \u001b[32m5\u001b[39m) -> Self:\n\u001b[32m   5676\u001b[39m \u001b[38;5;250m    \u001b[39m\u001b[33;03m\"\"\"\u001b[39;00m\n\u001b[32m   5677\u001b[39m \u001b[33;03m    Return the first `n` rows.\u001b[39;00m\n\u001b[32m   5678\u001b[39m \n\u001b[32m   (...)\u001b[39m\u001b[32m   5759\u001b[39m \u001b[33;03m    5     parrot\u001b[39;00m\n\u001b[32m   5760\u001b[39m \u001b[33;03m    \"\"\"\u001b[39;00m\n\u001b[32m-> \u001b[39m\u001b[32m5761\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43miloc\u001b[49m\u001b[43m[\u001b[49m\u001b[43m:\u001b[49m\u001b[43mn\u001b[49m\u001b[43m]\u001b[49m.copy()\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/miniforge3/envs/scanpy_env/lib/python3.12/site-packages/pandas/core/indexing.py:1207\u001b[39m, in \u001b[36m_LocationIndexer.__getitem__\u001b[39m\u001b[34m(self, key)\u001b[39m\n\u001b[32m   1205\u001b[39m maybe_callable = com.apply_if_callable(key, \u001b[38;5;28mself\u001b[39m.obj)\n\u001b[32m   1206\u001b[39m maybe_callable = \u001b[38;5;28mself\u001b[39m._raise_callable_usage(key, maybe_callable)\n\u001b[32m-> \u001b[39m\u001b[32m1207\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_getitem_axis\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmaybe_callable\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis\u001b[49m\u001b[43m=\u001b[49m\u001b[43maxis\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/miniforge3/envs/scanpy_env/lib/python3.12/site-packages/pandas/core/indexing.py:1750\u001b[39m, in \u001b[36m_iLocIndexer._getitem_axis\u001b[39m\u001b[34m(self, key, axis)\u001b[39m\n\u001b[32m   1744\u001b[39m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mIndexError\u001b[39;00m(\n\u001b[32m   1745\u001b[39m         \u001b[33m\"\u001b[39m\u001b[33mDataFrame indexer is not allowed for .iloc\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[33m\"\u001b[39m\n\u001b[32m   1746\u001b[39m         \u001b[33m\"\u001b[39m\u001b[33mConsider using .loc for automatic alignment.\u001b[39m\u001b[33m\"\u001b[39m\n\u001b[32m   1747\u001b[39m     )\n\u001b[32m   1749\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(key, \u001b[38;5;28mslice\u001b[39m):\n\u001b[32m-> \u001b[39m\u001b[32m1750\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_get_slice_axis\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis\u001b[49m\u001b[43m=\u001b[49m\u001b[43maxis\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m   1752\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m is_iterator(key):\n\u001b[32m   1753\u001b[39m     key = \u001b[38;5;28mlist\u001b[39m(key)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/miniforge3/envs/scanpy_env/lib/python3.12/site-packages/pandas/core/indexing.py:1785\u001b[39m, in \u001b[36m_iLocIndexer._get_slice_axis\u001b[39m\u001b[34m(self, slice_obj, axis)\u001b[39m\n\u001b[32m   1782\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m obj.copy(deep=\u001b[38;5;28;01mFalse\u001b[39;00m)\n\u001b[32m   1784\u001b[39m labels = obj._get_axis(axis)\n\u001b[32m-> \u001b[39m\u001b[32m1785\u001b[39m \u001b[43mlabels\u001b[49m\u001b[43m.\u001b[49m\u001b[43m_validate_positional_slice\u001b[49m\u001b[43m(\u001b[49m\u001b[43mslice_obj\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m   1786\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m.obj._slice(slice_obj, axis=axis)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/miniforge3/envs/scanpy_env/lib/python3.12/site-packages/pandas/core/indexes/base.py:4041\u001b[39m, in \u001b[36mIndex._validate_positional_slice\u001b[39m\u001b[34m(self, key)\u001b[39m\n\u001b[32m   4036\u001b[39m \u001b[38;5;250m\u001b[39m\u001b[33;03m\"\"\"\u001b[39;00m\n\u001b[32m   4037\u001b[39m \u001b[33;03mFor positional indexing, a slice must have either int or None\u001b[39;00m\n\u001b[32m   4038\u001b[39m \u001b[33;03mfor each of start, stop, and step.\u001b[39;00m\n\u001b[32m   4039\u001b[39m \u001b[33;03m\"\"\"\u001b[39;00m\n\u001b[32m   4040\u001b[39m \u001b[38;5;28mself\u001b[39m._validate_indexer(\u001b[33m\"\u001b[39m\u001b[33mpositional\u001b[39m\u001b[33m\"\u001b[39m, key.start, \u001b[33m\"\u001b[39m\u001b[33miloc\u001b[39m\u001b[33m\"\u001b[39m)\n\u001b[32m-> \u001b[39m\u001b[32m4041\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_validate_indexer\u001b[49m\u001b[43m(\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mpositional\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkey\u001b[49m\u001b[43m.\u001b[49m\u001b[43mstop\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43miloc\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[32m   4042\u001b[39m \u001b[38;5;28mself\u001b[39m._validate_indexer(\u001b[33m\"\u001b[39m\u001b[33mpositional\u001b[39m\u001b[33m\"\u001b[39m, key.step, \u001b[33m\"\u001b[39m\u001b[33miloc\u001b[39m\u001b[33m\"\u001b[39m)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/miniforge3/envs/scanpy_env/lib/python3.12/site-packages/pandas/core/indexes/base.py:6864\u001b[39m, in \u001b[36mIndex._validate_indexer\u001b[39m\u001b[34m(self, form, key, kind)\u001b[39m\n\u001b[32m   6859\u001b[39m \u001b[38;5;250m\u001b[39m\u001b[33;03m\"\"\"\u001b[39;00m\n\u001b[32m   6860\u001b[39m \u001b[33;03mIf we are positional indexer, validate that we have appropriate\u001b[39;00m\n\u001b[32m   6861\u001b[39m \u001b[33;03mtyped bounds must be an integer.\u001b[39;00m\n\u001b[32m   6862\u001b[39m \u001b[33;03m\"\"\"\u001b[39;00m\n\u001b[32m   6863\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m lib.is_int_or_none(key):\n\u001b[32m-> \u001b[39m\u001b[32m6864\u001b[39m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_raise_invalid_indexer\u001b[49m\u001b[43m(\u001b[49m\u001b[43mform\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/miniforge3/envs/scanpy_env/lib/python3.12/site-packages/pandas/core/indexes/base.py:4125\u001b[39m, in \u001b[36mIndex._raise_invalid_indexer\u001b[39m\u001b[34m(self, form, key, reraise)\u001b[39m\n\u001b[32m   4123\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m reraise \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m lib.no_default:\n\u001b[32m   4124\u001b[39m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m(msg) \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01mreraise\u001b[39;00m\n\u001b[32m-> \u001b[39m\u001b[32m4125\u001b[39m \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m(msg)\n",
      "\u001b[31mTypeError\u001b[39m: cannot do positional indexing on Index with these indexers [50] of type str"
     ]
    }
   ],
   "source": [
    "\n",
    "# 过滤细胞\n",
    "sc.pp.filter_cells(adata, min_genes=200)\n",
    "sc.pp.filter_genes(adata, min_cells=3)\n",
    "adata = adata[adata.obs.pct_counts_mt < 5, :] # 过滤线粒体占比高的细胞\n",
    "adata.obs.head('50')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f048f281",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 标准化与高变基因选择\n",
    "sc.pp.normalize_total(adata, target_sum=1e4)\n",
    "sc.pp.log1p(adata)\n",
    "sc.pp.highly_variable_genes(adata, min_mean=0.0125, max_mean=3, min_disp=0.5)\n",
    "adata.obs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56eb82e5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a5fead6c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0828bf13",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b8c5d8a3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53e0ac53",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ddd6addd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd070ac2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7e4013c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bfcab04b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63adce33",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ca88eab",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe9cef60",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a831f6a0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e3294022",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c76d0530",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56556d4c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scanpy_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
